xds/internal/xdsclient/filter_chain.go

```diff
@@ -65,6 +67,96 @@ type FilterChain struct {
 	//
 	// Only one of RouteConfigName and InlineRouteConfig is set.
 	InlineRouteConfig *RouteConfigUpdate
+
+	// InstantiatedFilters represents whether the xds client data has been
+	// converted to something usable for routing. Since this usable data will be
+	// used for any accepted connections that match to this filter chain, it is
+	// best to persist here.
+	InstantiatedFilters bool // Any mutexes?
+	// VirtualHosts are the virtual hosts ready to be used in the xds interceptors.
+	// It contains a way to match routes using a matcher and also instantiates
+	// HTTPFilter overrides to simply run incoming RPC's through if they are selected.
+	VirtualHosts []VirtualHostWithInterceptors
+}
+
+// VirtualHostWithInterceptors captures information present in a VirtualHost
+// update, and also contains routes with instantiated HTTP Filters.
+type VirtualHostWithInterceptors struct {
+	// Domains are the domain names which map to this Virtual Host. On the
+	// server side, this will be dictated by the :authority header of the
+	// incoming RPC.
+	Domains []string
+	// Routes are the Routes for this Virtual Host.
+	Routes []RouteWithInterceptors
+}
+
+// RouteWithInterceptors captures information in a Route, and contains
+// a usable matcher and also instantiated HTTP Filters.
+type RouteWithInterceptors struct {
+	// M is the matcher used to match to this route.
+	M *CompositeMatcher
+	// Interceptors are interceptors instantiated for this route. These will be
+	// constructed from a combination of the top level configuration and any
+	// HTTP Filter overrides present in Virtual Host or Route.
+	Interceptors []resolver.ServerInterceptor
+}
+
+// ConstructUsableRouteConfiguration uses the HTTP Filters persisted and Route
+// Configuration from Listener (since configuration may come from RDS) and
+// converts it into usable route configuration, including Instantiating any HTTP
+// Filters for each Route.
+func (f *FilterChain) ConstructUsableRouteConfiguration(config RouteConfigUpdate) error {
+	if f.InstantiatedFilters {
+		return nil
+	}
+	f.InstantiatedFilters = true
```

Comment:
Can you explain why this bool is necessary?  Is there a reason we might call this multiple times with the same inputs?  Or if we are calling it with different inputs, why is it okay to ignore subsequent invocations with data that would lead to different results?

Reply:
Ugh, I've been thinking about this a lot. There doesn't seem to good way of caching anything, so I went ahead and deleted this conditional. Some options I think might work are if Inline configuration cache it, or if received list of rds updates in lis wrapper ping all the filter chains to switch this back to false. But it seems like these are too complicated for the cache value? Idk, I went ahead and deleted it, let me know if you think one of those two or if you think of another potential way of caching.